src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -555,6 +553,264 @@ static void ucc_tl_ucp_ctx_remote_pack_data(ucc_tl_ucp_context_t *ctx,
     }
 }
 
+ucc_status_t ucc_tl_ucp_mem_map_memhbuf(ucc_tl_ucp_context_t *ctx,
+                                        void *pack_buffer, ucp_mem_h *mh)
+{
+    ucp_mem_map_params_t mmap_params;
+    ucs_status_t         status;
+    void                *packed_memh;
+
+    *mh = NULL;
+    /* unpack here */
+    packed_memh = UCC_TL_UCP_MEMH_TL_PACKED_MEMH(pack_buffer);
+
+    mmap_params.field_mask = UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER;
+    mmap_params.exported_memh_buffer = packed_memh;
+
+    status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, mh);
+    if (UCS_OK != status) {
+        if (status != UCS_ERR_UNREACHABLE) {
+            tl_error(ctx->super.super.lib,
+                     "ucp_mem_map failed with error code: %s",
+                     ucs_status_string(status));
+            return ucs_status_to_ucc_status(status);
+        } else {
+            tl_warn(ctx->super.super.lib,
+                     "ucp_mem_map cannot map exported memory handles");
+        }
+    }
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_mem_map_export(ucc_tl_ucp_context_t *ctx, void *address,
+                                       size_t len,
+                                       int type,
+                                       ucc_tl_ucp_memh_data_t *m_data)
+{
+    ucc_status_t           ucc_status = UCC_OK;
+    ucp_mem_h              mh;
+    ucp_mem_map_params_t   mmap_params;
+    ucp_memh_pack_params_t pack_params;
+    ucs_status_t           status;
+
+    if (type == UCC_MEM_MAP_TYPE_LOCAL) {
+        mmap_params.field_mask =
+            UCP_MEM_MAP_PARAM_FIELD_ADDRESS | UCP_MEM_MAP_PARAM_FIELD_LENGTH;
+        mmap_params.address = address;
+        mmap_params.length  = len;
+
+        status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+        if (UCS_OK != status) {
+            tl_error(ctx->super.super.lib, "ucp_mem_map failed with error code: %d",
+                     status);
+            ucc_status = ucs_status_to_ucc_status(status);
+            return ucc_status;
+        }
+        m_data->rinfo.mem_h   = mh;
+        m_data->rinfo.va_base = address;
+        m_data->rinfo.len     = len;
+
+        pack_params.field_mask = UCP_MEMH_PACK_PARAM_FIELD_FLAGS;
+        pack_params.flags      = UCP_MEMH_PACK_FLAG_EXPORT;
+
+        status = ucp_memh_pack(mh, &pack_params, &m_data->packed_memh,
```

Comment:
why do we do ucp_memh_pack and rkey_pack in mem_map function and not in memh_pack function?

Reply:
Primarily for convenience. I've updated this so it's correct and the ucp_memh_pack and rkey_pack are in the memh_pack function now.